libsplinter/src/service/scabbard/client/transaction.rs

```diff
@@ -0,0 +1,453 @@
+                compute_org_address(org_id)?,
+                compute_agent_address(&public_key)?,
+            ];
+
+            (addresses.clone(), addresses)
+        }
+        Action::DeleteSmartPermission(delete_smart_permission) => {
+            let org_id = delete_smart_permission.org_id();
+            let name = delete_smart_permission.name();
+            let addresses = vec![
+                compute_smart_permission_address(org_id, name)?,
+                compute_org_address(org_id)?,
+                compute_agent_address(&public_key)?,
+            ];
+
+            (addresses.clone(), addresses)
+        }
+    };
+
+    txn_header.set_inputs(protobuf::RepeatedField::from_vec(input_addresses));
+    txn_header.set_outputs(protobuf::RepeatedField::from_vec(output_addresses));
+
+    let payload_bytes = payload.into_bytes()?;
+    let hash = HashSigner
+        .sign(&payload_bytes)
+        .map_err(|err| Error::new_with_source("failed to hash payload bytes", err.into()))?;
+    txn_header.set_payload_sha512(bytes_to_hex_str(&hash));
+    txn.set_payload(payload_bytes);
+
+    let txn_header_bytes = txn_header.write_to_bytes()?;
+    txn.set_header(txn_header_bytes.clone());
+
+    let b: &[u8] = &txn_header_bytes;
+    txn.set_header_signature(signer.sign(b)?);
+
+    Ok(txn)
+}
+
+/// Returns a Batch for the given Transactions and Signer
+///
+/// # Arguments
+///
+/// * `txns` - transaction to put in the batch
+/// * `signer` - the signer to be used to sign the batch
+///
+/// # Errors
+///
+/// An error is returned if it occurs during serialization of the provided transactions or
+/// internally created `BatchHeader`, or if a signing error occurs.
+pub fn create_batch(txns: Vec<Transaction>, signer: &Signer) -> Result<Batch, Error> {
+    let public_key = signer
+        .get_public_key()
+        .map_err(|err| Error::new_with_source("failed to get public key from signer", err.into()))?
+        .as_hex();
+
+    let mut batch = Batch::new();
+    let mut batch_header = BatchHeader::new();
+
+    batch_header.set_transaction_ids(protobuf::RepeatedField::from_iter(
+        txns.iter().map(|txn| txn.header_signature.clone()),
+    ));
+    batch_header.set_signer_public_key(public_key);
+    batch.set_transactions(txns.into());
+
+    let batch_header_bytes = batch_header.write_to_bytes()?;
+    batch.set_header(batch_header_bytes.clone());
+
+    let b: &[u8] = &batch_header_bytes;
+    batch.set_header_signature(signer.sign(b)?);
+
+    Ok(batch)
+}
+
+/// Returns a BatchList containing the provided Batch
+///
+/// # Arguments
+///
+/// * `batch` - a Batch
+pub fn create_batch_list_from_one(batch: Batch) -> BatchList {
+    let mut batch_list = BatchList::new();
+    batch_list.set_batches(protobuf::RepeatedField::from_vec(vec![batch]));
+    batch_list
+}
+
+/// Creates a nonce appropriate for a TransactionHeader
+fn create_nonce() -> String {
+    let elapsed = Instant::now().elapsed();
```

Comment:
Isn't this always going to be a very small number - it only gives you the amount of time since the `Instant` was created?  I think you might be looking more for a system time.

Reply:
Removed from this PR